server/aesdsocket.c

```diff
@@ -158,146 +137,351 @@ int main(int argc, char* argv[])
+	{
+		if(ptr->connectionInProgress == false)
+		{
+			// This thread can be removed
+			pthread_join(ptr->threadId, NULL);
+			totalConnections--;
+			SLIST_REMOVE(&head, ptr, threadInfo_s, entries);
+			free(ptr);
+		}
+	}
+}
+
+/*
+ * Function for handling alarm signal
+ */
+void AlarmHandler(int signo)
+{
+	hasTimeElapsed = true;
+	CleanUpList();
+	if(totalConnections == 0)
+	{
+		int fd = fd = open(AESD_SOCKET_DATA_FILE, O_RDWR | O_APPEND | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
 		if(fd == -1)
 		{
 			perror("socket data file open");
-			close(acceptedSocketfd);
-			freeaddrinfo(serverInfo);
-			close(serverSockfd);
-			return -1;
+			exit(-1);
 		}
+		WriteTimestampToFile(fd);
+	}
+}
+
+void* ThreadSendAndReceive(void* threadParams)
+{
+	// Typecast the arguments
+	struct threadInfo_s* params = (struct threadInfo_s*)(threadParams);
+
+	// Log client IP to syslog
+	char clientIpStr[INET6_ADDRSTRLEN];
+	inet_ntop(AF_INET, params->clientInfo, clientIpStr, sizeof(clientIpStr));
+	syslog(LOG_INFO, "Accept connection from %s", clientIpStr);
+
+	char* pktBuf = NULL;
+	int fd, rc;
+	int currentBufSize = PACKET_BUFFER_SIZE;
+	ssize_t retVal;
+	char* ptr;
+	int remainingSize = 0;
 	
-		// Receive data from client and append to data file
-		pktBuf = (char*)(malloc(PACKET_BUFFER_SIZE * sizeof(char))); // Packet buffer of 512 chars
-		int currentBufSize = PACKET_BUFFER_SIZE;
-		ssize_t retVal;
-		char* ptr;
-		int remainingSize = 0;
+	// Receive data from client and append to data file
+	pktBuf = (char*)(malloc(PACKET_BUFFER_SIZE * sizeof(char))); // Packet buffer of 512 chars
+	if(pktBuf == NULL)
+	{
+		printf("Error while creating packet buffer!\n");
+		close(fd);
+		close(params->sockFd);
+		exit(-1);
+	}
 	
-		if(pktBuf == NULL)
+	memset(pktBuf, 0, currentBufSize); // Make all zeroes
+	
+	while((retVal = recv(params->sockFd, pktBuf + currentBufSize - PACKET_BUFFER_SIZE, PACKET_BUFFER_SIZE, 0)) != 0) // Until nothing more to read
+	{
+		if(retVal == -1)
 		{
-			printf("Error while creating packet buffer!\n");
+			perror("recv");
+			free(pktBuf);
 			close(fd);
-			close(acceptedSocketfd);
-			freeaddrinfo(serverInfo);
-			close(serverSockfd);
-			return -1;
+			close(params->sockFd);
+			exit(-1);
 		}
-	
-		memset(pktBuf, 0, currentBufSize); // Make all zeroes
-	
-		while((retVal = recv(acceptedSocketfd, pktBuf + currentBufSize - PACKET_BUFFER_SIZE, PACKET_BUFFER_SIZE, 0)) != 0) // Until nothing more to read
+
+		// Lock mutex
+		rc = pthread_mutex_lock(&mutex);
+		if(rc != 0)
 		{
-			if(retVal == -1)
+			printf("Error while obtaining mutex: %d", rc);
+			exit(-1);
+		}
+
+		// Open the file
+		fd = open(AESD_SOCKET_DATA_FILE, O_RDWR | O_APPEND | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+		if(fd == -1)
+		{
+			perror("socket data file open");
+			close(params->sockFd);
+			exit(-1);
+		}
+
+		// Check if time has elapsed
+		if(hasTimeElapsed)
+		{
+			WriteTimestampToFile(fd);
```

Comment:
@suryakanteti I think it will be tricky to add timestamp writes both here and in the signal handler. See my other comments about this.  Think about a design where you could handle from one place.